{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9fae68bd-a4a6-48e2-a511-f0972eafa7a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5a82afd-3fe3-4d5f-83a8-52793dd8f99f",
   "metadata": {},
   "source": [
    "# 1.DATOS BULK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "149f0525-32bb-4ee1-8813-de716c8329bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "bulk_datos=\"/home/franciscojavier/Escritorio/TFG/scDEAL/data/ALL_expression.csv\"\n",
    "Bulk= pd.read_csv(bulk_datos, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "aac71637-1120-46e0-b6ed-1ce020d068cb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ST3GAL2</th>\n",
       "      <th>NOTCH2</th>\n",
       "      <th>GRPEL1</th>\n",
       "      <th>SHC2</th>\n",
       "      <th>MAGEA8</th>\n",
       "      <th>GNA14</th>\n",
       "      <th>KRTCAP2</th>\n",
       "      <th>CEP44</th>\n",
       "      <th>EHD2</th>\n",
       "      <th>UTP14C</th>\n",
       "      <th>...</th>\n",
       "      <th>CIAPIN1</th>\n",
       "      <th>PPFIA4</th>\n",
       "      <th>EPPK1</th>\n",
       "      <th>PSPC1</th>\n",
       "      <th>SLC30A2</th>\n",
       "      <th>RNF13</th>\n",
       "      <th>LAMC1</th>\n",
       "      <th>PITX3</th>\n",
       "      <th>SNX24</th>\n",
       "      <th>IL17RE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DATA.683665</th>\n",
       "      <td>5.144679</td>\n",
       "      <td>3.223748</td>\n",
       "      <td>8.473993</td>\n",
       "      <td>3.367309</td>\n",
       "      <td>3.228205</td>\n",
       "      <td>2.980778</td>\n",
       "      <td>8.523072</td>\n",
       "      <td>3.568470</td>\n",
       "      <td>3.205451</td>\n",
       "      <td>6.173123</td>\n",
       "      <td>...</td>\n",
       "      <td>5.794856</td>\n",
       "      <td>3.115977</td>\n",
       "      <td>3.236775</td>\n",
       "      <td>4.614050</td>\n",
       "      <td>3.334627</td>\n",
       "      <td>8.512252</td>\n",
       "      <td>5.454001</td>\n",
       "      <td>3.579842</td>\n",
       "      <td>3.719559</td>\n",
       "      <td>3.208217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATA.683667</th>\n",
       "      <td>4.891309</td>\n",
       "      <td>4.978066</td>\n",
       "      <td>8.965003</td>\n",
       "      <td>7.597621</td>\n",
       "      <td>2.939119</td>\n",
       "      <td>3.127610</td>\n",
       "      <td>9.164316</td>\n",
       "      <td>3.226157</td>\n",
       "      <td>3.594308</td>\n",
       "      <td>4.838035</td>\n",
       "      <td>...</td>\n",
       "      <td>5.429998</td>\n",
       "      <td>3.165675</td>\n",
       "      <td>3.293903</td>\n",
       "      <td>3.833966</td>\n",
       "      <td>3.115850</td>\n",
       "      <td>9.187912</td>\n",
       "      <td>7.297341</td>\n",
       "      <td>3.350470</td>\n",
       "      <td>6.678025</td>\n",
       "      <td>3.106149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATA.684052</th>\n",
       "      <td>4.185672</td>\n",
       "      <td>4.580091</td>\n",
       "      <td>8.342951</td>\n",
       "      <td>3.454799</td>\n",
       "      <td>2.940054</td>\n",
       "      <td>3.842441</td>\n",
       "      <td>8.715775</td>\n",
       "      <td>3.509050</td>\n",
       "      <td>2.943156</td>\n",
       "      <td>5.688459</td>\n",
       "      <td>...</td>\n",
       "      <td>6.166544</td>\n",
       "      <td>3.070921</td>\n",
       "      <td>4.455511</td>\n",
       "      <td>4.108411</td>\n",
       "      <td>3.080420</td>\n",
       "      <td>7.986220</td>\n",
       "      <td>7.260711</td>\n",
       "      <td>3.123496</td>\n",
       "      <td>4.627073</td>\n",
       "      <td>3.088657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATA.684057</th>\n",
       "      <td>4.103474</td>\n",
       "      <td>4.681022</td>\n",
       "      <td>7.894892</td>\n",
       "      <td>6.939582</td>\n",
       "      <td>3.098473</td>\n",
       "      <td>4.450950</td>\n",
       "      <td>9.803396</td>\n",
       "      <td>3.732709</td>\n",
       "      <td>2.991773</td>\n",
       "      <td>6.523652</td>\n",
       "      <td>...</td>\n",
       "      <td>5.617250</td>\n",
       "      <td>3.318546</td>\n",
       "      <td>8.393024</td>\n",
       "      <td>3.945185</td>\n",
       "      <td>3.074119</td>\n",
       "      <td>8.615635</td>\n",
       "      <td>6.793222</td>\n",
       "      <td>3.177603</td>\n",
       "      <td>4.703124</td>\n",
       "      <td>3.081756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATA.684059</th>\n",
       "      <td>4.564867</td>\n",
       "      <td>4.076783</td>\n",
       "      <td>8.136921</td>\n",
       "      <td>7.620565</td>\n",
       "      <td>3.111922</td>\n",
       "      <td>6.328582</td>\n",
       "      <td>8.801057</td>\n",
       "      <td>3.633442</td>\n",
       "      <td>3.264081</td>\n",
       "      <td>6.073701</td>\n",
       "      <td>...</td>\n",
       "      <td>5.948226</td>\n",
       "      <td>3.203330</td>\n",
       "      <td>3.278578</td>\n",
       "      <td>3.672100</td>\n",
       "      <td>3.171120</td>\n",
       "      <td>8.673449</td>\n",
       "      <td>6.673377</td>\n",
       "      <td>3.496695</td>\n",
       "      <td>5.829292</td>\n",
       "      <td>3.210553</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 15962 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              ST3GAL2    NOTCH2    GRPEL1      SHC2    MAGEA8     GNA14  \\\n",
       "DATA.683665  5.144679  3.223748  8.473993  3.367309  3.228205  2.980778   \n",
       "DATA.683667  4.891309  4.978066  8.965003  7.597621  2.939119  3.127610   \n",
       "DATA.684052  4.185672  4.580091  8.342951  3.454799  2.940054  3.842441   \n",
       "DATA.684057  4.103474  4.681022  7.894892  6.939582  3.098473  4.450950   \n",
       "DATA.684059  4.564867  4.076783  8.136921  7.620565  3.111922  6.328582   \n",
       "\n",
       "              KRTCAP2     CEP44      EHD2    UTP14C  ...   CIAPIN1    PPFIA4  \\\n",
       "DATA.683665  8.523072  3.568470  3.205451  6.173123  ...  5.794856  3.115977   \n",
       "DATA.683667  9.164316  3.226157  3.594308  4.838035  ...  5.429998  3.165675   \n",
       "DATA.684052  8.715775  3.509050  2.943156  5.688459  ...  6.166544  3.070921   \n",
       "DATA.684057  9.803396  3.732709  2.991773  6.523652  ...  5.617250  3.318546   \n",
       "DATA.684059  8.801057  3.633442  3.264081  6.073701  ...  5.948226  3.203330   \n",
       "\n",
       "                EPPK1     PSPC1   SLC30A2     RNF13     LAMC1     PITX3  \\\n",
       "DATA.683665  3.236775  4.614050  3.334627  8.512252  5.454001  3.579842   \n",
       "DATA.683667  3.293903  3.833966  3.115850  9.187912  7.297341  3.350470   \n",
       "DATA.684052  4.455511  4.108411  3.080420  7.986220  7.260711  3.123496   \n",
       "DATA.684057  8.393024  3.945185  3.074119  8.615635  6.793222  3.177603   \n",
       "DATA.684059  3.278578  3.672100  3.171120  8.673449  6.673377  3.496695   \n",
       "\n",
       "                SNX24    IL17RE  \n",
       "DATA.683665  3.719559  3.208217  \n",
       "DATA.683667  6.678025  3.106149  \n",
       "DATA.684052  4.627073  3.088657  \n",
       "DATA.684057  4.703124  3.081756  \n",
       "DATA.684059  5.829292  3.210553  \n",
       "\n",
       "[5 rows x 15962 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Bulk.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "caf02dd0-c2b2-4521-af30-4a42a204bce5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "El archivo tiene 1280 filas y 15962 columnas.\n"
     ]
    }
   ],
   "source": [
    "print(f\"El archivo tiene {Bulk.shape[0]} filas y {Bulk.shape[1]} columnas.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b5ab5910-7c93-45a7-9b87-dc7b92e48730",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Número de genes: 15962\n",
      "Número de muestras: 1280\n",
      "Primeros 5 genes: ['ST3GAL2', 'NOTCH2', 'GRPEL1', 'SHC2', 'MAGEA8']\n"
     ]
    }
   ],
   "source": [
    "print(f\"Número de genes: {Bulk.shape[1]}\")\n",
    "print(f\"Número de muestras: {Bulk.shape[0]}\")\n",
    "print(\"Primeros 5 genes:\", Bulk.columns[:5].tolist())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35983961-ddb9-4eb0-b239-5a04b07df979",
   "metadata": {},
   "source": [
    "# 2.CALCULAMOS VARIANZA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6d8cdbda-e02d-4207-bd37-f31f8d4b0e84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ST3GAL2    0.892848\n",
      "NOTCH2     0.813878\n",
      "GRPEL1     2.395445\n",
      "SHC2       3.767519\n",
      "MAGEA8     2.661780\n",
      "             ...   \n",
      "RNF13      4.454409\n",
      "LAMC1      2.325935\n",
      "PITX3      2.773513\n",
      "SNX24      1.557141\n",
      "IL17RE     1.291969\n",
      "Length: 15962, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "varianzas = Bulk.var(axis=0)  # varianza por columna (gen)\n",
    "print(varianzas)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "07e35229-7b3a-4467-9312-d77ae3e4a58c",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_20_porcentaje = int(len(varianzas) * 0.2) #Filtramos el top 20\n",
    "genes_top = varianzas.sort_values(ascending=False).head(top_20_porcentaje).index\n",
    "df_top_genes = Bulk[genes_top]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "34397ebd-007a-4e4b-b8c4-c9b08a64c797",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes seleccionados: 3192 de 15962 originales\n"
     ]
    }
   ],
   "source": [
    "print(f\"Genes seleccionados: {df_top_genes.shape[1]} de {Bulk.shape[1]} originales\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5b807a02-778c-4baf-aa5e-9fc423118124",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RPS4Y1</th>\n",
       "      <th>KRT19</th>\n",
       "      <th>VIM</th>\n",
       "      <th>S100P</th>\n",
       "      <th>EEF1A1</th>\n",
       "      <th>RPL18A</th>\n",
       "      <th>TACSTD2</th>\n",
       "      <th>RPS18</th>\n",
       "      <th>BASP1</th>\n",
       "      <th>RPS19</th>\n",
       "      <th>...</th>\n",
       "      <th>RAET1G</th>\n",
       "      <th>IGFL3</th>\n",
       "      <th>ANGPT1</th>\n",
       "      <th>RRP1B</th>\n",
       "      <th>C2ORF42</th>\n",
       "      <th>IFNGR1</th>\n",
       "      <th>TCEAL5</th>\n",
       "      <th>DUSP11</th>\n",
       "      <th>GNA11</th>\n",
       "      <th>MFSD6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DATA.683665</th>\n",
       "      <td>10.744227</td>\n",
       "      <td>3.405941</td>\n",
       "      <td>12.085272</td>\n",
       "      <td>3.336260</td>\n",
       "      <td>4.922896</td>\n",
       "      <td>3.062681</td>\n",
       "      <td>3.496972</td>\n",
       "      <td>4.855941</td>\n",
       "      <td>5.697826</td>\n",
       "      <td>4.038713</td>\n",
       "      <td>...</td>\n",
       "      <td>4.280225</td>\n",
       "      <td>3.394853</td>\n",
       "      <td>3.002283</td>\n",
       "      <td>8.001636</td>\n",
       "      <td>6.736740</td>\n",
       "      <td>7.831898</td>\n",
       "      <td>3.023507</td>\n",
       "      <td>7.876978</td>\n",
       "      <td>4.013298</td>\n",
       "      <td>7.270762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATA.683667</th>\n",
       "      <td>12.410156</td>\n",
       "      <td>4.226230</td>\n",
       "      <td>11.661481</td>\n",
       "      <td>5.922569</td>\n",
       "      <td>6.533475</td>\n",
       "      <td>3.113093</td>\n",
       "      <td>3.027207</td>\n",
       "      <td>4.680126</td>\n",
       "      <td>5.011485</td>\n",
       "      <td>4.045782</td>\n",
       "      <td>...</td>\n",
       "      <td>3.581134</td>\n",
       "      <td>3.379538</td>\n",
       "      <td>3.001898</td>\n",
       "      <td>8.308363</td>\n",
       "      <td>6.997331</td>\n",
       "      <td>7.443246</td>\n",
       "      <td>3.019110</td>\n",
       "      <td>7.700416</td>\n",
       "      <td>8.309896</td>\n",
       "      <td>4.358078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATA.684052</th>\n",
       "      <td>3.374015</td>\n",
       "      <td>4.032691</td>\n",
       "      <td>12.671484</td>\n",
       "      <td>3.477540</td>\n",
       "      <td>5.551361</td>\n",
       "      <td>2.892388</td>\n",
       "      <td>2.832211</td>\n",
       "      <td>3.884602</td>\n",
       "      <td>11.453684</td>\n",
       "      <td>3.944376</td>\n",
       "      <td>...</td>\n",
       "      <td>3.144730</td>\n",
       "      <td>3.307964</td>\n",
       "      <td>3.387063</td>\n",
       "      <td>7.957953</td>\n",
       "      <td>6.356331</td>\n",
       "      <td>6.905754</td>\n",
       "      <td>3.405001</td>\n",
       "      <td>7.464873</td>\n",
       "      <td>8.508063</td>\n",
       "      <td>4.455841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATA.684057</th>\n",
       "      <td>12.822256</td>\n",
       "      <td>3.091820</td>\n",
       "      <td>10.667924</td>\n",
       "      <td>3.595975</td>\n",
       "      <td>5.277746</td>\n",
       "      <td>2.932124</td>\n",
       "      <td>2.932878</td>\n",
       "      <td>4.880253</td>\n",
       "      <td>9.359950</td>\n",
       "      <td>3.950865</td>\n",
       "      <td>...</td>\n",
       "      <td>3.840384</td>\n",
       "      <td>3.390363</td>\n",
       "      <td>2.994032</td>\n",
       "      <td>8.219939</td>\n",
       "      <td>6.465303</td>\n",
       "      <td>8.016595</td>\n",
       "      <td>4.280432</td>\n",
       "      <td>7.801002</td>\n",
       "      <td>7.849750</td>\n",
       "      <td>3.848845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATA.684059</th>\n",
       "      <td>3.320986</td>\n",
       "      <td>3.492342</td>\n",
       "      <td>12.870585</td>\n",
       "      <td>4.567399</td>\n",
       "      <td>5.819835</td>\n",
       "      <td>2.982949</td>\n",
       "      <td>2.923305</td>\n",
       "      <td>3.888214</td>\n",
       "      <td>10.924775</td>\n",
       "      <td>3.717019</td>\n",
       "      <td>...</td>\n",
       "      <td>3.285166</td>\n",
       "      <td>3.518197</td>\n",
       "      <td>3.736578</td>\n",
       "      <td>7.958470</td>\n",
       "      <td>6.950287</td>\n",
       "      <td>8.002485</td>\n",
       "      <td>3.412617</td>\n",
       "      <td>6.972215</td>\n",
       "      <td>8.140381</td>\n",
       "      <td>4.290012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3192 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                RPS4Y1     KRT19        VIM     S100P    EEF1A1    RPL18A  \\\n",
       "DATA.683665  10.744227  3.405941  12.085272  3.336260  4.922896  3.062681   \n",
       "DATA.683667  12.410156  4.226230  11.661481  5.922569  6.533475  3.113093   \n",
       "DATA.684052   3.374015  4.032691  12.671484  3.477540  5.551361  2.892388   \n",
       "DATA.684057  12.822256  3.091820  10.667924  3.595975  5.277746  2.932124   \n",
       "DATA.684059   3.320986  3.492342  12.870585  4.567399  5.819835  2.982949   \n",
       "\n",
       "              TACSTD2     RPS18      BASP1     RPS19  ...    RAET1G     IGFL3  \\\n",
       "DATA.683665  3.496972  4.855941   5.697826  4.038713  ...  4.280225  3.394853   \n",
       "DATA.683667  3.027207  4.680126   5.011485  4.045782  ...  3.581134  3.379538   \n",
       "DATA.684052  2.832211  3.884602  11.453684  3.944376  ...  3.144730  3.307964   \n",
       "DATA.684057  2.932878  4.880253   9.359950  3.950865  ...  3.840384  3.390363   \n",
       "DATA.684059  2.923305  3.888214  10.924775  3.717019  ...  3.285166  3.518197   \n",
       "\n",
       "               ANGPT1     RRP1B   C2ORF42    IFNGR1    TCEAL5    DUSP11  \\\n",
       "DATA.683665  3.002283  8.001636  6.736740  7.831898  3.023507  7.876978   \n",
       "DATA.683667  3.001898  8.308363  6.997331  7.443246  3.019110  7.700416   \n",
       "DATA.684052  3.387063  7.957953  6.356331  6.905754  3.405001  7.464873   \n",
       "DATA.684057  2.994032  8.219939  6.465303  8.016595  4.280432  7.801002   \n",
       "DATA.684059  3.736578  7.958470  6.950287  8.002485  3.412617  6.972215   \n",
       "\n",
       "                GNA11     MFSD6  \n",
       "DATA.683665  4.013298  7.270762  \n",
       "DATA.683667  8.309896  4.358078  \n",
       "DATA.684052  8.508063  4.455841  \n",
       "DATA.684057  7.849750  3.848845  \n",
       "DATA.684059  8.140381  4.290012  \n",
       "\n",
       "[5 rows x 3192 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_top_genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a809793-3083-43eb-bc76-258617872443",
   "metadata": {},
   "source": [
    "# 3.MAPEAR A RATÓN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "22bb908e-fc07-4148-bb06-a9936b4f7583",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pip install mygene\n",
    "import mygene\n",
    "# Lista de genes humanos que quieres mapear \n",
    "genes_humanos = list(df_top_genes.columns)  # por ejemplo si tu DataFrame ya está filtrado\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e58addae-6b6c-4a3f-990c-40469e9b28c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Input sequence provided is already in string format. No operation performed\n",
      "Input sequence provided is already in string format. No operation performed\n",
      "5 input query terms found dup hits:\t[('EPPK1', 2), ('ST6GALNAC2', 2), ('MAGEF1', 2), ('NNT', 2), ('LCN15', 2)]\n",
      "305 input query terms found no hit:\t['RPS4Y1', 'S100P', 'C19ORF33', 'CLEC2B', 'HNRNPA1L2', 'HLA-DPA1', 'HLA-DRA', 'AKR1C1', 'AKR1C3', 'A\n"
     ]
    }
   ],
   "source": [
    "#Inicializamos\n",
    "mg = mygene.MyGeneInfo()\n",
    "# Buscar ortólogos en ratón\n",
    "resultados = mg.querymany(genes_humanos, scopes='symbol', fields='ensembl.gene', species='mouse')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "387f51e9-b941-494f-9cc8-514b9c77a1d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapeo = {}\n",
    "for r in resultados: #MApeamos solo aquellos q tiene 1 solo homologo en raton , los demas los descartamos que son los de la celda de arriba\n",
    "    if 'notfound' not in r and 'ensembl' in r:\n",
    "        # Puede haber uno o varios resultados (lista o dict)\n",
    "        if isinstance(r['ensembl'], list):\n",
    "            mapeo[r['query']] = r['ensembl'][0]['gene']\n",
    "        else:\n",
    "            mapeo[r['query']] = r['ensembl']['gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1f9efda4-dc92-4260-9423-2ebbdf5a2400",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes mapeados: 2883 de 3192\n"
     ]
    }
   ],
   "source": [
    "print(f\"Genes mapeados: {len(mapeo)} de {len(genes_humanos)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "68515c5b-104f-473a-b91c-f5cfd20354aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KRT19  →  ENSMUSG00000020911\n",
      "VIM  →  ENSMUSG00000026728\n",
      "EEF1A1  →  ENSMUSG00000037742\n",
      "RPL18A  →  ENSMUSG00000045128\n",
      "TACSTD2  →  ENSMUSG00000051397\n",
      "RPS18  →  ENSMUSG00000008668\n",
      "BASP1  →  ENSMUSG00000045763\n",
      "RPS19  →  ENSMUSG00000040952\n",
      "SRGN  →  ENSMUSG00000020077\n",
      "UCHL1  →  ENSMUSG00000029223\n"
     ]
    }
   ],
   "source": [
    "for humano, raton in list(mapeo.items())[:10]:\n",
    "    print(f\"{humano}  →  {raton}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e0186daa-1ed5-4d67-87c0-e41d35ffd710",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_raton_bulk = list(mapeo.values())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59bb6f9f-29d4-46d2-a65b-dd6a423d24db",
   "metadata": {},
   "source": [
    "# SINGLE CELL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ef0d2804-da03-4fd9-9251-2bace4349626",
   "metadata": {},
   "outputs": [],
   "source": [
    "ruta_single_cell=\"/home/franciscojavier/Escritorio/TFG/scDEAL/data/GSE110894/GSE110894.csv\"\n",
    "Single_cell= pd.read_csv(ruta_single_cell,index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "767093d1-aef4-4e01-98d6-f8e56eade1cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RPI1_A3</th>\n",
       "      <th>RPI1_A4</th>\n",
       "      <th>RPI1_A5</th>\n",
       "      <th>RPI1_A6</th>\n",
       "      <th>RPI1_A7</th>\n",
       "      <th>RPI1_A8</th>\n",
       "      <th>RPI1_A9</th>\n",
       "      <th>RPI1_A10</th>\n",
       "      <th>RPI1_A11</th>\n",
       "      <th>RPI1_A12</th>\n",
       "      <th>...</th>\n",
       "      <th>RPI7_P13</th>\n",
       "      <th>RPI7_P14</th>\n",
       "      <th>RPI7_P15</th>\n",
       "      <th>RPI7_P16</th>\n",
       "      <th>RPI7_P17</th>\n",
       "      <th>RPI7_P18</th>\n",
       "      <th>RPI7_P19</th>\n",
       "      <th>RPI7_P20</th>\n",
       "      <th>RPI7_P21</th>\n",
       "      <th>RPI7_P22</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000110331</th>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000006281</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000040726</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000024829</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000029007</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1504 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    RPI1_A3  RPI1_A4  RPI1_A5  RPI1_A6  RPI1_A7  RPI1_A8  \\\n",
       "gene_id                                                                    \n",
       "ENSMUSG00000110331        3        5        5        6        5        2   \n",
       "ENSMUSG00000006281        0        0        0        1        0        0   \n",
       "ENSMUSG00000040726        0        0        0        0        0        0   \n",
       "ENSMUSG00000024829        1        2        0        0        1        3   \n",
       "ENSMUSG00000029007        1        2        1        0        0        0   \n",
       "\n",
       "                    RPI1_A9  RPI1_A10  RPI1_A11  RPI1_A12  ...  RPI7_P13  \\\n",
       "gene_id                                                    ...             \n",
       "ENSMUSG00000110331        4        10        10         1  ...         0   \n",
       "ENSMUSG00000006281        0         0         0         1  ...         0   \n",
       "ENSMUSG00000040726        0         0         0         0  ...         0   \n",
       "ENSMUSG00000024829        1         1         3         1  ...         0   \n",
       "ENSMUSG00000029007        0         0         0         0  ...         0   \n",
       "\n",
       "                    RPI7_P14  RPI7_P15  RPI7_P16  RPI7_P17  RPI7_P18  \\\n",
       "gene_id                                                                \n",
       "ENSMUSG00000110331         6         5         1         5         4   \n",
       "ENSMUSG00000006281         1         0         0         0         1   \n",
       "ENSMUSG00000040726         0         0         0         0         0   \n",
       "ENSMUSG00000024829         0         1         0         1         1   \n",
       "ENSMUSG00000029007         0         0         0         0         0   \n",
       "\n",
       "                    RPI7_P19  RPI7_P20  RPI7_P21  RPI7_P22  \n",
       "gene_id                                                     \n",
       "ENSMUSG00000110331         3         3         0         4  \n",
       "ENSMUSG00000006281         0         0         0         0  \n",
       "ENSMUSG00000040726         0         1         0         0  \n",
       "ENSMUSG00000024829         2         0         0         1  \n",
       "ENSMUSG00000029007         0         1         0         0  \n",
       "\n",
       "[5 rows x 1504 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Single_cell.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "3580fe0b-97b3-46d0-b281-bd4688595d6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "El archivo tiene 25921 filas y 1504 columnas.\n"
     ]
    }
   ],
   "source": [
    "print(f\"El archivo tiene {Single_cell.shape[0]} filas y {Single_cell.shape[1]} columnas.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "24085d63-2fd3-45fa-978e-e09661417f21",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_raton_sc = list(Single_cell.index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "d47c6101-5b48-4f5e-91d3-c5c6cfea5867",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_comunes = list(set(genes_raton_bulk) & set(genes_raton_sc))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "a4b976ed-d671-416a-882e-e134d9fd8921",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes comunes encontrados: 2100\n"
     ]
    }
   ],
   "source": [
    "print(f\"Genes comunes encontrados: {len(genes_comunes)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "eb957427-3851-4193-ae70-3e5b446fa924",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_comunes.sort()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a29c16df-673e-478a-98eb-d10e2e053384",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7e63e3d-8d47-460e-8f2d-ff726f672058",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "740d916e-d6d5-4c3a-b68f-2bf1d32a8d20",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5b20d5ec-c90e-47af-8e4f-811434cdcca0",
   "metadata": {},
   "source": [
    "# FILTRAMOS DATOS BULK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "f0ae4b7b-bc88-4573-a06a-d3b9e784ddb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapeo_inverso = {v: k for k, v in mapeo.items()}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "e8ff2dc9-f7f9-4862-8b42-cdde676608cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_comunes.sort()\n",
    "genes_humanos_ordenados = [mapeo_inverso[ensm] for ensm in genes_comunes if ensm in mapeo_inverso]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "2067e94c-34ba-4299-9412-5f658b22061c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_bulk_humano_ordenado = Bulk[genes_humanos_ordenados]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "2c9351ff-b76d-4f4d-9a5b-f1c000396397",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GNAI3</th>\n",
       "      <th>APOH</th>\n",
       "      <th>CAV2</th>\n",
       "      <th>TBX2</th>\n",
       "      <th>SLC22A18</th>\n",
       "      <th>CCND2</th>\n",
       "      <th>GMPR</th>\n",
       "      <th>ITGB2</th>\n",
       "      <th>CDH4</th>\n",
       "      <th>MX1</th>\n",
       "      <th>...</th>\n",
       "      <th>CDKN2D</th>\n",
       "      <th>PSMB9</th>\n",
       "      <th>ZFP82</th>\n",
       "      <th>BIN2</th>\n",
       "      <th>SOWAHC</th>\n",
       "      <th>ARL14</th>\n",
       "      <th>KCTD12</th>\n",
       "      <th>LHB</th>\n",
       "      <th>PRMT1</th>\n",
       "      <th>IGIP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DATA.683665</th>\n",
       "      <td>8.418405</td>\n",
       "      <td>3.108346</td>\n",
       "      <td>3.202542</td>\n",
       "      <td>3.169560</td>\n",
       "      <td>3.657305</td>\n",
       "      <td>9.506000</td>\n",
       "      <td>3.019949</td>\n",
       "      <td>8.579302</td>\n",
       "      <td>4.090406</td>\n",
       "      <td>10.763135</td>\n",
       "      <td>...</td>\n",
       "      <td>5.942791</td>\n",
       "      <td>10.719714</td>\n",
       "      <td>5.733338</td>\n",
       "      <td>5.779168</td>\n",
       "      <td>2.960625</td>\n",
       "      <td>2.875492</td>\n",
       "      <td>3.263250</td>\n",
       "      <td>5.264080</td>\n",
       "      <td>4.677651</td>\n",
       "      <td>5.697826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATA.683667</th>\n",
       "      <td>8.384469</td>\n",
       "      <td>3.157720</td>\n",
       "      <td>3.288323</td>\n",
       "      <td>3.250875</td>\n",
       "      <td>7.147396</td>\n",
       "      <td>8.888872</td>\n",
       "      <td>7.729681</td>\n",
       "      <td>3.410079</td>\n",
       "      <td>9.935629</td>\n",
       "      <td>2.872464</td>\n",
       "      <td>...</td>\n",
       "      <td>7.881910</td>\n",
       "      <td>5.660218</td>\n",
       "      <td>6.856965</td>\n",
       "      <td>3.054499</td>\n",
       "      <td>6.515636</td>\n",
       "      <td>2.961698</td>\n",
       "      <td>3.204516</td>\n",
       "      <td>4.882622</td>\n",
       "      <td>4.497861</td>\n",
       "      <td>6.156337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATA.684052</th>\n",
       "      <td>8.860598</td>\n",
       "      <td>3.867216</td>\n",
       "      <td>11.630509</td>\n",
       "      <td>2.811466</td>\n",
       "      <td>6.257794</td>\n",
       "      <td>3.107823</td>\n",
       "      <td>3.541739</td>\n",
       "      <td>5.445975</td>\n",
       "      <td>5.639173</td>\n",
       "      <td>2.948423</td>\n",
       "      <td>...</td>\n",
       "      <td>6.110756</td>\n",
       "      <td>5.967821</td>\n",
       "      <td>6.320678</td>\n",
       "      <td>3.113162</td>\n",
       "      <td>6.973525</td>\n",
       "      <td>3.102612</td>\n",
       "      <td>3.167568</td>\n",
       "      <td>4.527282</td>\n",
       "      <td>4.477633</td>\n",
       "      <td>5.892460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATA.684057</th>\n",
       "      <td>9.405991</td>\n",
       "      <td>5.940434</td>\n",
       "      <td>9.510553</td>\n",
       "      <td>3.025210</td>\n",
       "      <td>3.619437</td>\n",
       "      <td>5.924566</td>\n",
       "      <td>5.540122</td>\n",
       "      <td>4.157251</td>\n",
       "      <td>5.349462</td>\n",
       "      <td>3.007122</td>\n",
       "      <td>...</td>\n",
       "      <td>6.656717</td>\n",
       "      <td>4.057070</td>\n",
       "      <td>7.095704</td>\n",
       "      <td>3.066957</td>\n",
       "      <td>7.150156</td>\n",
       "      <td>2.913979</td>\n",
       "      <td>4.024446</td>\n",
       "      <td>4.069318</td>\n",
       "      <td>4.508922</td>\n",
       "      <td>7.254817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DATA.684059</th>\n",
       "      <td>8.871912</td>\n",
       "      <td>3.145452</td>\n",
       "      <td>10.057207</td>\n",
       "      <td>3.168384</td>\n",
       "      <td>5.456578</td>\n",
       "      <td>7.344971</td>\n",
       "      <td>5.913958</td>\n",
       "      <td>3.740122</td>\n",
       "      <td>3.314965</td>\n",
       "      <td>5.795148</td>\n",
       "      <td>...</td>\n",
       "      <td>5.347125</td>\n",
       "      <td>7.250015</td>\n",
       "      <td>6.617563</td>\n",
       "      <td>2.883226</td>\n",
       "      <td>7.244060</td>\n",
       "      <td>3.004077</td>\n",
       "      <td>3.409099</td>\n",
       "      <td>4.134250</td>\n",
       "      <td>4.488666</td>\n",
       "      <td>6.882576</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 2100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GNAI3      APOH       CAV2      TBX2  SLC22A18     CCND2  \\\n",
       "DATA.683665  8.418405  3.108346   3.202542  3.169560  3.657305  9.506000   \n",
       "DATA.683667  8.384469  3.157720   3.288323  3.250875  7.147396  8.888872   \n",
       "DATA.684052  8.860598  3.867216  11.630509  2.811466  6.257794  3.107823   \n",
       "DATA.684057  9.405991  5.940434   9.510553  3.025210  3.619437  5.924566   \n",
       "DATA.684059  8.871912  3.145452  10.057207  3.168384  5.456578  7.344971   \n",
       "\n",
       "                 GMPR     ITGB2      CDH4        MX1  ...    CDKN2D  \\\n",
       "DATA.683665  3.019949  8.579302  4.090406  10.763135  ...  5.942791   \n",
       "DATA.683667  7.729681  3.410079  9.935629   2.872464  ...  7.881910   \n",
       "DATA.684052  3.541739  5.445975  5.639173   2.948423  ...  6.110756   \n",
       "DATA.684057  5.540122  4.157251  5.349462   3.007122  ...  6.656717   \n",
       "DATA.684059  5.913958  3.740122  3.314965   5.795148  ...  5.347125   \n",
       "\n",
       "                 PSMB9     ZFP82      BIN2    SOWAHC     ARL14    KCTD12  \\\n",
       "DATA.683665  10.719714  5.733338  5.779168  2.960625  2.875492  3.263250   \n",
       "DATA.683667   5.660218  6.856965  3.054499  6.515636  2.961698  3.204516   \n",
       "DATA.684052   5.967821  6.320678  3.113162  6.973525  3.102612  3.167568   \n",
       "DATA.684057   4.057070  7.095704  3.066957  7.150156  2.913979  4.024446   \n",
       "DATA.684059   7.250015  6.617563  2.883226  7.244060  3.004077  3.409099   \n",
       "\n",
       "                  LHB     PRMT1      IGIP  \n",
       "DATA.683665  5.264080  4.677651  5.697826  \n",
       "DATA.683667  4.882622  4.497861  6.156337  \n",
       "DATA.684052  4.527282  4.477633  5.892460  \n",
       "DATA.684057  4.069318  4.508922  7.254817  \n",
       "DATA.684059  4.134250  4.488666  6.882576  \n",
       "\n",
       "[5 rows x 2100 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bulk_humano_ordenado.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "0a840468-e95e-4293-b721-38d10c93ab96",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_bulk_humano_ordenado.to_csv(\"bulk_2100_humano.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "af07149f-801f-4029-9530-6a440e7a8314",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "labels = pd.read_csv(\"/home/franciscojavier/Escritorio/TFG/scDEAL/data/ALL_label_binary_wf.csv\", index_col=0)\n",
    "print((labels.index == df_bulk_humano_ordenado.index).all())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b5f1d78-44e4-428d-af9e-4c5e9dd40ab0",
   "metadata": {},
   "source": [
    "# FILTRAMOS DATOS SINGLE CELL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "d1776696-3504-415b-8608-e75ad7e5f374",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_comunes.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "0577a8b3-1944-4aec-955b-29eeb78085ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sc_final = Single_cell.loc[genes_comunes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "6ac9a0e5-39cb-4511-98b2-e6d968e64e2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RPI1_A3</th>\n",
       "      <th>RPI1_A4</th>\n",
       "      <th>RPI1_A5</th>\n",
       "      <th>RPI1_A6</th>\n",
       "      <th>RPI1_A7</th>\n",
       "      <th>RPI1_A8</th>\n",
       "      <th>RPI1_A9</th>\n",
       "      <th>RPI1_A10</th>\n",
       "      <th>RPI1_A11</th>\n",
       "      <th>RPI1_A12</th>\n",
       "      <th>...</th>\n",
       "      <th>RPI7_P13</th>\n",
       "      <th>RPI7_P14</th>\n",
       "      <th>RPI7_P15</th>\n",
       "      <th>RPI7_P16</th>\n",
       "      <th>RPI7_P17</th>\n",
       "      <th>RPI7_P18</th>\n",
       "      <th>RPI7_P19</th>\n",
       "      <th>RPI7_P20</th>\n",
       "      <th>RPI7_P21</th>\n",
       "      <th>RPI7_P22</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000049</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000058</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000093</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000154</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1504 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    RPI1_A3  RPI1_A4  RPI1_A5  RPI1_A6  RPI1_A7  RPI1_A8  \\\n",
       "gene_id                                                                    \n",
       "ENSMUSG00000000001        3        8        5       11        7        1   \n",
       "ENSMUSG00000000049        0        0        0        0        0        0   \n",
       "ENSMUSG00000000058        0        0        0        0        0        0   \n",
       "ENSMUSG00000000093        0        0        0        0        0        0   \n",
       "ENSMUSG00000000154        0        0        0        0        0        0   \n",
       "\n",
       "                    RPI1_A9  RPI1_A10  RPI1_A11  RPI1_A12  ...  RPI7_P13  \\\n",
       "gene_id                                                    ...             \n",
       "ENSMUSG00000000001        4         4         3         1  ...         0   \n",
       "ENSMUSG00000000049        0         0         0         0  ...         0   \n",
       "ENSMUSG00000000058        0         0         0         0  ...         0   \n",
       "ENSMUSG00000000093        0         0         0         0  ...         0   \n",
       "ENSMUSG00000000154        0         0         0         0  ...         0   \n",
       "\n",
       "                    RPI7_P14  RPI7_P15  RPI7_P16  RPI7_P17  RPI7_P18  \\\n",
       "gene_id                                                                \n",
       "ENSMUSG00000000001         4         7         0         2         5   \n",
       "ENSMUSG00000000049         0         0         0         0         0   \n",
       "ENSMUSG00000000058         0         0         0         0         1   \n",
       "ENSMUSG00000000093         0         0         0         0         0   \n",
       "ENSMUSG00000000154         0         0         0         0         0   \n",
       "\n",
       "                    RPI7_P19  RPI7_P20  RPI7_P21  RPI7_P22  \n",
       "gene_id                                                     \n",
       "ENSMUSG00000000001         4         1         0         2  \n",
       "ENSMUSG00000000049         0         0         0         0  \n",
       "ENSMUSG00000000058         0         0         0         0  \n",
       "ENSMUSG00000000093         0         0         0         0  \n",
       "ENSMUSG00000000154         0         0         0         0  \n",
       "\n",
       "[5 rows x 1504 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sc_final.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "45a574d8-3de1-43f2-8008-d646091fe0f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sc_final.to_csv(\"sc_2100genes.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "7dbb8159-418e-4042-a312-37844bee0997",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n"
     ]
    }
   ],
   "source": [
    "# Cargar el bulk que usaste para entrenar\n",
    "df_bulk = pd.read_csv(\"bulk_2100_humano.csv\", index_col=0)\n",
    "# Verificar orden\n",
    "print(list(df_bulk_humano_ordenado.columns) == list(df_sc_final.index))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30203a2b-e261-42e8-8b55-bc7492f3a3b8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "7d62ec2a-103b-48fe-a680-f9483fd5d202",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Your filename has more than two extensions: ['.BET', '.762_bottle_512_edim_256,128_pdim_128,64_model_DAETOP2_dropout_0', '.1_gene_F_lr_0', '.5_mod_new_sam_upsampling', '.h5ad'].\n",
      "Only considering the two last: ['.5_mod_new_sam_upsampling', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.BET', '.762_bottle_512_edim_256,128_pdim_128,64_model_DAETOP2_dropout_0', '.1_gene_F_lr_0', '.5_mod_new_sam_upsampling', '.h5ad'].\n",
      "Only considering the two last: ['.5_mod_new_sam_upsampling', '.h5ad'].\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "adata = sc.read(\"/home/franciscojavier/Escritorio/TFG/scDEAL/save/adata/GSE110894_integrate_data_GSE110894_drug_I.BET.762_bottle_512_edim_256,128_pdim_128,64_model_DAETOP2_dropout_0.1_gene_F_lr_0.5_mod_new_sam_upsampling.h5ad\")  # reemplaza <para> por el sufijo real del modelo\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "167e0bf4-29f0-4ae3-8da2-4d0652adec5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "y_true = adata.obs['sensitive']             # Etiqueta verdadera\n",
    "y_pred = adata.obs['sens_label'].astype(int)  # Etiqueta predicha por el modelo\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "d86e8fbb-eff0-4bf0-b8c0-923b0764b650",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.9513742071881607\n",
      "F1 Score: 0.9513884124211166\n",
      "Precision: 0.9518193478096653\n",
      "Recall: 0.9513742071881607\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.94      0.96      0.95       685\n",
      "           1       0.97      0.94      0.95       734\n",
      "\n",
      "    accuracy                           0.95      1419\n",
      "   macro avg       0.95      0.95      0.95      1419\n",
      "weighted avg       0.95      0.95      0.95      1419\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import classification_report, accuracy_score, f1_score, precision_score, recall_score\n",
    "\n",
    "print(\"Accuracy:\", accuracy_score(y_true, y_pred))\n",
    "print(\"F1 Score:\", f1_score(y_true, y_pred, average='weighted'))\n",
    "print(\"Precision:\", precision_score(y_true, y_pred, average='weighted'))\n",
    "print(\"Recall:\", recall_score(y_true, y_pred, average='weighted'))\n",
    "\n",
    "# Reporte completo\n",
    "print(classification_report(y_true, y_pred))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "853c4672-f3d4-49c7-a8b0-bc9f2254a2ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Your filename has more than two extensions: ['.BET', '.762_bottle_512_edim_256,128_pdim_128,64_model_DAETOP2_dropout_0', '.1_gene_F_lr_0', '.5_mod_new_sam_upsampling', '.h5ad'].\n",
      "Only considering the two last: ['.5_mod_new_sam_upsampling', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.BET', '.762_bottle_512_edim_256,128_pdim_128,64_model_DAETOP2_dropout_0', '.1_gene_F_lr_0', '.5_mod_new_sam_upsampling', '.h5ad'].\n",
      "Only considering the two last: ['.5_mod_new_sam_upsampling', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.BET', '.762_bottle_512_edim_256,128_pdim_128,64_model_DAE_dropout_0', '.1_gene_F_lr_0', '.5_mod_new_sam_upsampling', '.h5ad'].\n",
      "Only considering the two last: ['.5_mod_new_sam_upsampling', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.BET', '.762_bottle_512_edim_256,128_pdim_128,64_model_DAE_dropout_0', '.1_gene_F_lr_0', '.5_mod_new_sam_upsampling', '.h5ad'].\n",
      "Only considering the two last: ['.5_mod_new_sam_upsampling', '.h5ad'].\n",
      "🎯 Comparación entre modelos\n",
      "Accuracy entre predicciones: 0.9796\n",
      "F1-score entre predicciones: 0.9796\n",
      "¿Son exactamente iguales? False\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "from sklearn.metrics import accuracy_score, f1_score\n",
    "\n",
    "# Cargar ambos archivos\n",
    "adata1 = sc.read(\"/home/franciscojavier/Escritorio/TFG/scDEAL/save/adata/GSE110894_integrate_data_GSE110894_drug_I.BET.762_bottle_512_edim_256,128_pdim_128,64_model_DAETOP2_dropout_0.1_gene_F_lr_0.5_mod_new_sam_upsampling.h5ad\")  # <-- reemplaza por el nombre real\n",
    "adata2 = sc.read(\"/home/franciscojavier/Escritorio/TFG/scDEAL/save/adata/GSE110894integrate_data_GSE110894_drug_I.BET.762_bottle_512_edim_256,128_pdim_128,64_model_DAE_dropout_0.1_gene_F_lr_0.5_mod_new_sam_upsampling.h5ad\")\n",
    "\n",
    "# Extraer predicciones\n",
    "pred1 = adata1.obs['sens_label'].astype(int)\n",
    "pred2 = adata2.obs['sens_label'].astype(int)\n",
    "\n",
    "# Verificar que las células coinciden\n",
    "assert all(adata1.obs.index == adata2.obs.index), \"Las muestras no están alineadas\"\n",
    "\n",
    "# Comparar métricamente\n",
    "accuracy = accuracy_score(pred1, pred2)\n",
    "f1 = f1_score(pred1, pred2, average='weighted')\n",
    "\n",
    "print(\"🎯 Comparación entre modelos\")\n",
    "print(f\"Accuracy entre predicciones: {accuracy:.4f}\")\n",
    "print(f\"F1-score entre predicciones: {f1:.4f}\")\n",
    "\n",
    "# (Opcional) ¿Son idénticos?\n",
    "print(\"¿Son exactamente iguales?\", np.array_equal(pred1, pred2))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (scDEALenv)",
   "language": "python",
   "name": "scdealenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
